File: src/main/scala/array/ValidMountainArray.scala
package array

/**
  * leetcode 941
  */
object ValidMountainArray {
  def validMountainArray(A: Array[Int]): Boolean = {
    if (A == null || A.length < 3) {
      return false
    }
    var max = (0, A(0))
    for (i <- 0 until A.length) {
      if (A(i) >= max._2) {
        max = (i, A(i))
      }
    }


    if (max._1 == A.length - 1 || max._1 == 0) return false

    var before = A.slice(0, max._1 + 1)
    var after = A.slice(max._1, A.length)


    var temp = before(0)
    for (i <- 1 until before.length) {
      if (temp >= before(i)) return false
      else temp = before(i)
    }
    temp = after(0)
    for (i <- 1 until after.length) {
      if (temp <= after(i)) return false
      else temp = after(i)
    }
    true
  }

  def main(args: Array[String]): Unit = {
    var arr = Array(14, 82, 89, 84, 79, 70, 70, 68, 67, 66, 63, 60, 58, 54, 44, 43, 32, 28, 26, 25, 22, 15, 13, 12, 10, 8, 7, 5, 4, 3)
    println(validMountainArray(arr))
  }
}

File: src/main/java/io/github/gcstar/leetcode/problem941/ValidMountainArray.java
package io.github.gcstar.leetcode.problem941;

import java.util.Arrays;

/**
 * Created by guchao on 19/1/7.
 */
public class ValidMountainArray {
    public boolean validMountainArray(int[] A) {
        if (A == null || A.length < 3) return false;
        int[] max = new int[]{0,A[0]};
        for (int i = 0; i < A.length; i++) {
            if (A[i] > max[1]) {
                max[0] = i;
                max[1] = A[i];
            }
        }
        if (max[0] == 0 || max[0] == A.length - 1) return false;
        int[] before = Arrays.copyOfRange(A, 0, max[0] + 1);
        int[] after = Arrays.copyOfRange(A, max[0], A.length);
        int temp = before[0];
        for (int i = 1; i < before.length; i++) {
            if (before[i] <= temp) return false;
            temp = before[i];
        }
        temp = after[0];
        for (int i = 1; i < after.length; i++) {
            if (after[i]>= temp) return false;
            temp = after[i];
        }
        return true;
    }
}
